### ISSUE TITLE
Dataframe model outputs?

### ISSUE THREAD

**Issue by joerunde** [SeldonIO/MLServer#416]
Yo @adriangonz!

I had previously added support for dataframe input to SKLearn runtimes here: https://github.com/SeldonIO/MLServer/pull/379

Our users are supper happy with that 👍 , and now we've been asked about sort of the opposite: supporting dataframes as an output from an SKLearn model.

We have users with something like this as the last stage of an SKLearn pipeline:
```
class SuperModelThing():
...
    def predict(self, X):
        # Some cool prediction thing
        ...
        output = pandas.DataFrame()
        output["object"] = "Apple"
        output["object_probability"] = 0.79
        output["is_vegetable"] = 1
        output["is_vegetable_probability"] = 0.42
        return output
```

The SKLearn runtime doesn't currently support encoding a response from a data frame, and it's made more awkward by the fact that it also assumes each item in the outputs is a response from calling one function on the model. 
I get that fits the normal usage paradigm- you have a model that returns an array of labels from `predict()` and an array of probabilities of those labels from `predict_proba()`, so I'm wondering how best to proceed here.

1. We could ask the user to adapt their code to use `predict` and `predict_proba` as expected...
```
class SuperModelThing():
...
    def predict(self, X):
        ...
        object = "Apple"
        is_vegetable = 1
        
        return np.array([
            class_name_to_int(object),
            is_vegetable
        ])

        
    def predict_proba(self, X):
        object_proba = 0.79
        is_vegetable_proba = 0.42

        return np.array([
            object_proba,
            is_vegetable_proba
        ])
```
But then they'd be responsible for more mapping on the caller side, they'd need to understand which index in the output corresponds to which feature and undoing any transformations like going from class name -> integer code -> class name.

2. We could ask the user to expose each output feature as a callable function on the model:
```
class SuperModelThing():
...
    def object(self, X):
        ...

    def object_proba(self, X):
        ...

    def is_vegetable(self, X):
        ...

    def is_vegetable_proba(self, X):
        ...
```
...But then I'm not quite sure how you'd go about actually implementing that in a stateless and threadsafe manner without re-running the underlying models for each output, which seems like a non-starter.

3. We could modify the SKLearn runtime to include some oddball custom logic like:
```
if len(payload.outputs == 1) and (payload.outputs[0].name = PREDICT_OUTPUT):
    output = self.model.predict(X)
    if isinstance(output, pd.DataFrame):
        return PandasCodec.encode(output)
```
But then of course you have a model where you request the `predict` output and get a response with a bunch of differently named outputs, and you break a ton of assumptions in the SKLearn runtime.

We hadn't really thought through this particular use case yet, so I'm wondering if this has come up before either here in MLServer or elsewhere in the KServe community. Any chance you have some thoughts or pointers on how to proceed?
Do you think the answer here is to conform to the assumptions that the SKLearn runtime makes, use a different MLServer runtime, or use a different runtime altogether?

**Comment by joerunde** [comment]
> Would something like that be a "user-friendly-enough" solution ? 

Yeah, if we're not super tied to an assumption that the named outputs in the request must map to the named outputs in the response, and that if we get one dataframe back out of whatever model method we call then we just encode that... that doesn't seem too bad.

We've gotten some code working internally for our users, I put up a PR here: https://github.com/SeldonIO/MLServer/pull/425

I'm still planning on making that a little less ugly and adding more test coverage and better warnings, but we'd love your opinion in the mean time

**Comment by adriangonz** [comment]
Yeah, I agree that the "requested outputs" bit may be a bit awkward. If we want to make it slightly more strict we can also follow something like:

1. If a user doesn't ask for any output, then
1.1. If the output is a plain tensor, then it would be sent back as `predict`.
1.2 If the output is a dataframe, then it would be a multi-output payload representing the table.
2. If a user explicitly asks for `predict` (or `predict_proba`), then
2.1 If the output is a plain tensor, then it's all good - same behaviour as right now (except the case where `predict_proba` is not supported, which is not currently handled explicitly anyway).
2.2 If the output is a dataframe, then we raise a warning (or an explicit error?), saying that the requested outputs are not available since it's a columnar output?

**Comment by joerunde** [comment]
Yeah, those slightly stricter rules could make sense. I think the only functional difference between that and what we have coded up now is that if the user asks for one of `predict` or `predict_proba` and the output is a data frame, then we'll return a multi-output payload without warning. We do throw an error if both are requested and at least one output is a data frame, since we wouldn't be able to jam the outputs together into one response. So I think we currently have:

1. If a user doesn't ask for any output, then
  1.1. If the output is a plain tensor, then it would be sent back as predict.
  1.2 If the output is a dataframe, then it would be a multi-output payload representing the table.
2. If a user explicitly asks for one of `predict` or `predict_proba`, then
  2.1 If the output is a plain tensor, then it's all good - same behaviour as right now
  2.2 If the output is a dataframe, then we return a multi-output payload representing the table.
3. If a user explicitly asks for both `predict` and `predict_proba`, then
  3.1 If both outputs are plain tensors, then the same behavior as right now
  3.2 If either output is a DataFrame, raise an error saying that the outputs are columnar data and can't be returned in a single request

For (2.2) I don't know if I really like the idea of raising an error, since we'll still be able to run the model and encode the data back into a response. But I could see how if a user explicitly requested an output and specified a different content type than `pd`, then technically we're not returning the data they asked for?
```
{
  "name": "predict",
  "parameters": {
    "content-type": "string"
  }
}
```
I suppose too if a user had a model with `predict_proba` returning a data frame, they would want a way to call it without it erroring? I could see a warning message being useful just to alert users that they shouldn't be looking for `predict` or `predict_proba` in their outputs, but at the same time it would probably get annoying to not have a way to run a successful request without generating a warning log.

Also
> except the case where predict_proba is not supported, which is not currently handled explicitly anyway

I think that actually is handled explicitly: https://github.com/SeldonIO/MLServer/blob/master/runtimes/sklearn/mlserver_sklearn/sklearn.py#L66-L70
